src/libexpr/package.nix

```diff
@@ -0,0 +1,134 @@
+{ lib
+, stdenv
+, releaseTools
+
+, meson
+, ninja
+, pkg-config
+, bison
+, flex
+, cmake # for resolving toml11 dep
+
+, nix-util
+, nix-store
+, nix-fetchers
+, boost
+, boehmgc
+, nlohmann_json
+, toml11
+
+# Configuration Options
+
+, versionSuffix ? ""
+
+# Check test coverage of Nix. Probably want to use with at least
+# one of `doCheck` or `doInstallCheck` enabled.
+, withCoverageChecks ? false
+
+# Whether to use garbage collection for the Nix language evaluator.
+#
+# If it is disabled, we just leak memory, but this is not as bad as it
+# sounds so long as evaluation just takes places within short-lived
+# processes. (When the process exits, the memory is reclaimed; it is
+# only leaked *within* the process.)
+#
+# Temporarily disabled on Windows because the `GC_throw_bad_alloc`
+# symbol is missing during linking.
+, enableGC ? !stdenv.hostPlatform.isWindows
+}:
+
+let
+  inherit (lib) fileset;
+
+  version = lib.fileContents ./.version + versionSuffix;
+
+  mkDerivation =
+    if withCoverageChecks
+    then
+      # TODO support `finalAttrs` args function in
+      # `releaseTools.coverageAnalysis`.
+      argsFun:
+         releaseTools.coverageAnalysis (let args = argsFun args; in args)
+    else stdenv.mkDerivation;
+in
+
+mkDerivation (finalAttrs: {
+  pname = "nix-expr";
+  inherit version;
+
+  src = fileset.toSource {
+    root = ./.;
+    fileset = fileset.unions [
+      ./meson.build
+      ./meson.options
+      ./primops/meson.build
+      (fileset.fileFilter (file: file.hasExt "cc") ./.)
+      (fileset.fileFilter (file: file.hasExt "hh") ./.)
+      ./lexer.l
+      ./parser.y
+      (fileset.fileFilter (file: file.hasExt "nix") ./.)
+    ];
+  };
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+    bison
+    flex
+    cmake
+  ];
+
+  buildInputs = [
+    toml11
+  ];
+
+  propagatedBuildInputs = [
+    nix-util
+    nix-store
+    nix-fetchers
+    boost
+    nlohmann_json
+  ] ++ lib.optional enableGC boehmgc;
+
+  preConfigure =
+    # "Inline" .version so it's not a symlink, and includes the suffix.
+    # Do the meson utils, without modification.
+    ''
+      echo ${version} > .version
+      cp -r ${../../build-utils-meson} build-utils-meson
```

Comment:
thought:
I'd previously proposed a fileset / `mkDerivation` / `setup.sh` feature where you'd just specify `../../build-utils-meson/something` in `fileset.unions` and it all just works. (And actually implemented it in its precursor, `sources.extend`, not merged)
Not actionable now, and while I think @infinisil was right to keep it simple first, I'd still like to have it :smiling_face_with_tear: .

Reply:
I guess you could emulate that solution though, by saying `root = ../..;` and then `cd`-ing into it in `preConfigure` or something. That way you don't have to mess around substituting symlinks, and we keep the packaged build a bit more similar to the local dev layout. It just wouldn't look quite as pretty as it could with the integrated subdirectory support.

Reply:
Yeah I really want a fileset feature to deref or rewrite a symlink. It would require some Nix-side changes to do the most nicely, I think, but that's fine.

Reply:
I think it's perhaps even useful to require that we declare an acceptable target, and in that case, we don't need any new built-in functionality.

Reply:
@roberth Err I mean that isn't everything in `fileset` implemented in terms of `builtins.filterSource` right now? I think we'd want something more flexible to create an arbitrary `SourceAccesor` with not just filtered out stuff, but replaced files too etc.

Reply:
Sure, but that's for a future Nix.

I highly highly recommend https://github.com/NixOS/nix/pull/10995.

Reply:
Yes, it is for future nix